### ISSUE TITLE
compiler: MoE decode ops — routed expert GEMV + router + shared expert + combine

### ISSUE THREAD

**Issue by xzyaoi** [opentela-ai/vkernels#98]
## Context
DeepSeek-V4/V4.1 MoE (`DeepseekV4SparseMoeBlock`, hash `tid2eid` router for the first `num_hash_layers`, learned top-k `sqrtsoftplus` router after; one replicated shared expert) and GLM-5.3 (`Glm53TopkRouter`/`Glm53Experts`, `swiglu_limit` clamping). Decode-side this is tractable with a **static task grid + runtime indirection**: k·B expert tasks, each reading its row’s routed expert weight base from a routing table written by the router phase.

## Semantics (decode step)
- router: small GEMV + top-k (static count) → routing table (expert-id + weight per (row, slot)) — external scratch
- experts: k·B `gemv_fp8`/bf16 gate-up+down tasks with **indirected weight base** (#94 pattern), `swiglu_limit` clamp folded into the swiglu task
- shared expert: plain dense path, adds like a residual
- combine: weighted scatter-add per row (one task per row)

## Work
- ops: `moe_route`, `expert_gemv` (indirected weights), `moe_combine`; lowering with static counts; legality: routing-table RAW orders router before experts before combine (phase order)
- MXFP4 expert weights: dtype follow-up on `gemv_fp8` (#91)

## Validation
eager MoE block on tiny configs (floe CPU tests) as oracle; routing-table determinism; bit-parity vs dense-every-expert evaluation at k=E.

Labels: arch:ops, model:deepseek-v4, model:glm-5.3

## Implementation sketch

**Backend**: Triton, static task grid + runtime indirection: router → `k·B` expert tasks (weight base read from the routing table) → one combine task per row. Phase barriers order router ≺ experts ≺ combine (RAW on the routing table).

**Arithmetic**:
```
router (fp32):   s = sigmoid(W_r·x)                          # noaux_tc
                 group = top-2 sums over expert groups → topk_group groups
                 idx  = top_k(s restricted to selected groups)  (sorted=False)
                 w    = renorm(s[idx]) if norm_topk_prob else s[idx]
hash layers:     idx = tid2eid[token_id]                      # frozen gather, k=1
expert slot e:   g,u  = W_g[e]·x, W_u[e]·x                     # bf16 or #91 fp8 GEMV, fp32 acc
                 act = silu(clamp(g, ≤L)) · clamp(u, ±L)       # swiglu_limit
                 h_e = W_d[e]·act
combine:         y   = Σ_k w_k · h_{idx_k} (+ shared expert dense path)
```
**Dtypes**: router logits fp32; expert weights bf16 initially, fp8/MXFP4 via #91 follow-up; routing table **i32 ids + fp32 weights** `[B, k]` scratch (external, phase-ordered); partials fp32, out bf16.

**Comment by xzyaoi** [comment]
## ISSUE 98 REPORT — moe decode ops vertical slice: complete, suite green

**Branch** `issue-98-moe-decode-ops` (pushed through `1163cf3`), PR opened against `main`. Bare-env suite: **24 passed, 2 skipped** (skips = floe eager-parity tests, import-gated per the bare-environment contract — attested, not verified). Full compiler lane after merging current `main` (#105/#106/#113/#115, and #94→#116 landed mid-flight): **361 passed**, with one pre-existing failure on `main` itself (`test_discovery` kernel-registry order — fails identically on `origin/main`, unrelated to this branch).

### Delivered (vertical slice, §4/§8/§9/§12 patterns)

1. **Capture** (`ops.moe_route` / `ops.moe_expert` / `ops.moe_combine`): the router writes the **external routing table** (i32 `ids` + f32 `weights`, `[B, k]`) that expert and combine phases consume through RAW hazards; the expert tasks do **runtime weight-base indirection** `e = ids[b, slot]` (the #94 slot-table pattern applied to a read-only weight pool); combine is the per-row weighted scatter-add `y = Σ_k w_k·h_k` (+ dense shared-expert row when recorded). Three router families: DeepSeek-V4 `sqrtsoftplus` global top-k, GLM-5.3 `sigmoid_noaux_tc` (bias + top-2-sum group restriction, unbiased score gather), and the frozen `hash` gather. Ties resolve to the lower expert index (stable descending) — documented determinism contract.
2. **Lowerings**: one route task per row, the static `k·B` expert grid (one task per (row, slot) computes that row's full expert FFN with `swiglu_limit` clamp folded in: `silu(min(g, L)) · clip(u, ±L)`), one combine task per row. Thread contract checked.
3. **Reference bodies**: fp64 oracles of the fp32 device math; `moe_route`/`moe_expert`/`moe_combine` dispatched in the reference executor.
4. **Triton template mirrors**: numpy mirrors of `_t_moe_route` / `_t_moe_expert` / `_t_moe_combine` decompositions (worker strides, f32 grids) — TEMPLATE_NAMES registered.
5. **Suite coverage** (all CPU): router determinism + tie rules (learned/noaux_tc/hash), hash exactness, **k=E bit-parity vs dense-every-expert** (weights rounded through the f32 table grid, f64 partials/accumulator — passes `assert_array_equal`), swiglu_limit clamp boundaries (pass-through at ±L, clamps beyond, negative gate unclamped), expert weight-base indirection (rewriting the router input re-points the weight rows), RAW phase ordering route ≺ experts ≺ combine, weighted combine vs mirror, shared-expert add, NaN canaries, one-launch accounting.

### Bugs found and fixed while validating (worth recording)

- **Mangled silu in the expert reference body**: `g/(1+sigmoid(g))·u` instead of `silu(g)·u` — caught by the clamp-boundary test only after the routing was made expert-distinct.
- **Mangled router sigmoid**: `1/(1+sigmoid(l))` instead of the stable `exp(-softplus(-l))`. Executor and test mirror agreed on the wrong formula, so route tests passed against a consistently wrong oracle — fixed on both sides to the floe form.
- **Workspace-reuse "ghost" symptoms**: post-run reads of dead fresh buffers show their successor's bytes (legal lifetime-disjoint co-allocation: `shared_down` reuses `shared_gate`'s region; the canary re-poisons buffers that die). Early debug traces that looked like NaN/aliasing corruption in the shared-expert chain were entirely this — the executor was correct; the debug reads were not.
- Test-harness fixes: k=E mirror now rounds through the executor's actual f32/f64 grids (bit-parity is real, not tolerance-purchased); clamp test redesigned with H=2 (with H=1 and positive x, the argmax is scale-invariant — rows can never hit distinct experts); weight-base test captures in hash mode and rewrites `token_ids`, since hash routing recomputes `route_ids` from the table.

### Notes

- The two stacked integration commits (#91 `gemv_fp8`, #94 slot-table indirection) are now ancestors of `main` (#115, #116); the branch merged current `main` cleanly — the PR stands alone.
- fp32 weights only in this slice; bf16/fp8-block expert variants are dtype follow-ups on the same task shape.
- `test_discovery` kernel-order failure reproduces on `origin/main` and predates this branch.

**Comment by xzyaoi** [comment]
Superseded by #129 (Stage 2 megakernel, merged): the MoE decode ops (moe_route / moe_expert / moe_combine) landed with the DeepSeek-V4 decode step. OP_MOE_ROUTE / OP_MOE_EXPERT / OP_MOE_COMBINE present on main. Closing per the supersession note on PR #123.
